(2nd) fix hook sync race when multiple set hooks command

Shrini-h opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
a. Alternate fix for the reverted https://github.com/openpbs/openpbs/pull/1947
b. two places where we are breaking instead of continuing while looping through all mom objects for hook related operations
1. https://github.com/openpbs/openpbs/blob/69d4e0258bd62541409305e3a0bd8b083dc52149/src/server/hook_func.c#L5072-L5075
2. https://github.com/openpbs/openpbs/blob/69d4e0258bd62541409305e3a0bd8b083dc52149/src/server/hook_func.c#L5161-L5164


When multiple set hook commands are issued from qmgr, some updates were not being propagated to execution nodes due to race in pending hook action processing

Bug is visible in the below screenshot
![InkedScreenshot_hook_unsynced_LI](https://user-images.githubusercontent.com/30307577/89184620-ba9f7d00-d5b6-11ea-85f5-ad95ffbd6cb3.jpg)
#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
fixed by 
1. using new hook action object when replies are pending for previous actions.
2. Introduced new flag member to `mom_hook_action_t` called `reply_expected` to track if we are waiting to hear back from mom.
3. Serialized mom hook sync operation triggers (`sync_mom_hookfilesTPP`)
4. Avoid marking hook sync pending during create node operation
5. simplified `add_pending_mom_allhooks_action()`
6. used direct unicast of batch request to delete mom hooks during node delete operation
7. renamed `bg_sync_mom_hookfiles()` to `mc_sync_mom_hookfiles()` to reflect the contemporary mechanism of hook syncs
8. renamed `bg_delete_mom_hooks()` to `uc_delete_mom_hooks()`

#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
[hook_sync_fix_2_logs.zip](https://github.com/openpbs/openpbs/files/5122788/hook_sync_fix_2_logs.zip)
![image](https://user-images.githubusercontent.com/30307577/91157845-7e59cb00-e6e3-11ea-8e26-51dc9ae05479.png)

Description: Tests from given sources on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|3354|3443|0|0|0|2|3441|

<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

subhasisb commented on src/server/hook_func.c:
```diff
@@ -5321,6 +5328,38 @@ mk_deferred_hook_info(int index, int event, long long int tid)
+int
+check_for_latest_action(mominfo_t *minfo, mom_hook_action_t *pact, int j, int event)
+{
+	/* if the same action is marked in the next actions for the same hook then remove
```
indentation

Shrini-h replied:
Done https://github.com/openpbs/openpbs/pull/1992/commits/d54ce17afb4e9d6cd1a3da070504726722e60e5c

subhasisb commented on src/server/hook_func.c:
```diff
@@ -5321,6 +5328,38 @@ mk_deferred_hook_info(int index, int event, long long int tid)
+	for (i = 0; i < minfo->mi_num_action; i++) {
+		pact2 = minfo->mi_action[i];
+		if (pact2 && (pact2->tid > pact->tid) && (pact2->action & event)
+				&& pact2->hookname && (strcmp(pact2->hookname, pact->hookname) == 0)) {
```
Better readability with the && in the previous line

Shrini-h replied:
Done https://github.com/openpbs/openpbs/pull/1992/commits/d54ce17afb4e9d6cd1a3da070504726722e60e5c

subhasisb commented on src/server/hook_func.c:
```diff
@@ -5321,6 +5328,38 @@ mk_deferred_hook_info(int index, int event, long long int tid)
+/**
+ * @brief
+ *		check if there is any new pending action
+ *		if yes remove the current one
```
does not look like this function itself does any removal

Shrini-h replied:
That was an initial design, Thanks for noticing. Now removed this line in https://github.com/openpbs/openpbs/pull/1992/commits/d54ce17afb4e9d6cd1a3da070504726722e60e5c

subhasisb commented on src/server/hook_func.c:
```diff
@@ -6012,128 +6063,124 @@ add_pending_mom_allhooks_action(void *minfo, unsigned int action)
-			phook->hook_name, action);
-		phook = (hook *)GET_NEXT(phook->hi_execjob_launch_hooks);
+		if (phook->hook_name &&	(phook->event & MOM_EVENTS)) {
+			add_pending_mom_hook_action((mominfo_t *)minfo,
```
merge to one line - these days we are leaning towards longer line lengths

Shrini-h replied:
Done https://github.com/openpbs/openpbs/pull/1992/commits/d54ce17afb4e9d6cd1a3da070504726722e60e5c

subhasisb commented on src/server/hook_func.c:
```diff
@@ -6012,128 +6063,124 @@ add_pending_mom_allhooks_action(void *minfo, unsigned int action)
+					snprintf(log_buffer, sizeof(log_buffer),
+						"timedout, clearing reply_expected for %d event[%lld] of %s hook for %s",
+						pact->reply_expected, tid, pact->hookname, pmom->mi_host);
+					log_event(PBSEVENT_DEBUG3, PBS_EVENTCLASS_SERVER,
```
log_eventf?

Shrini-h replied:
Done here and one more place down somewhere. https://github.com/openpbs/openpbs/pull/1992/commits/d54ce17afb4e9d6cd1a3da070504726722e60e5c

subhasisb commented on src/server/hook_func.c:
```diff
@@ -6012,128 +6063,124 @@ add_pending_mom_allhooks_action(void *minfo, unsigned int action)
  * @brief
- *		Checks to see if it's time to run bg_sync_mom_hookfiles() and if so,
- *		then run bg_sync_mom_hookfiles().
+ *		checks for hook sync operstion's timeout
```
operation spelling

Shrini-h replied:
done https://github.com/openpbs/openpbs/pull/1992/commits/d54ce17afb4e9d6cd1a3da070504726722e60e5c

subhasisb reviewed (changes requested):
Where are we making the send's serial such that if one reply is pending, then we would not send the next?

bayucan commented on src/include/hook_func.h:
```diff
@@ -71,6 +71,7 @@ extern "C" {
 struct mom_hook_action {
 	char            hookname[PBS_HOOK_NAME_SIZE];
 	unsigned int	action;
+	unsigned int	reply_expected; /* reply expected from mom for sent out actions */
```
Can you tell me exactly the values that this take? I scanned through server/hook_func.c, and I only see this set (or unset) as:
pact->reply_expected = 0U;
pact->reply_expected &= ~(event);

Shrini-h replied:
Thank you for flagging this. I must have inadvertently removed some lines while reverting some debug lines. Now I have carefully gone through my development branch and added back the missing lines in https://github.com/openpbs/openpbs/pull/1992/commits/d54ce17afb4e9d6cd1a3da070504726722e60e5c

bayucan commented on src/server/hook_func.c:
```diff
@@ -5321,6 +5328,38 @@ mk_deferred_hook_info(int index, int event, long long int tid)
+ * @retval 1  - found
+ */
+int
+check_for_latest_action(mominfo_t *minfo, mom_hook_action_t *pact, int j, int event)
```
I don't see the 'j' index being used at all in this function. Did you mean to skip checking the minfo entries that match the 'j' index, since that is actually 'pact'?

Shrini-h replied:
I was using it for a different purpose initially. and forgot to remove it later. But thanks, your suggestion is a good optimization. So fixed it in https://github.com/openpbs/openpbs/pull/1992/commits/d54ce17afb4e9d6cd1a3da070504726722e60e5c

bayucan replied:
Ok, thanks. I'm seeing it more clearly now.

bayucan commented on src/server/hook_func.c:
```diff
@@ -5535,6 +5581,12 @@ post_sendhookTPP(struct work_task *pwt)
+		pact->action &= ~(event);
+		hook_track_save(minfo, j);
+	}
+
```
What does the above do?

Shrini-h replied:
So when the server receives a failure reply for a sent hook action, it checks if there is a same action marked for pending for the same hook while we were waiting for the reply. If yes then it removes the action pending bit from the current action object. as it will be taken care in the next sync operation due to the new pending mark in a fresh action pending object and we want to avoid duplicating the pending mark.
Hope I got through.

bayucan replied:
Ok.

Shrini-h commented:
> Where are we making the send's serial such that if one reply is pending, then we would not send the next?

https://github.com/openpbs/openpbs/blob/d54ce17afb4e9d6cd1a3da070504726722e60e5c/src/server/hook_func.c#L6219

Shrini-h commented:
Thanks to @bayucan, I found out some line changes were missing after I squashed and cleaned my dev code. Now I have carefully gone through and added back the missing changes in https://github.com/openpbs/openpbs/pull/1992/commits/d54ce17afb4e9d6cd1a3da070504726722e60e5c

Shrini-h commented:
Note: this PR also fixes
b. two places where we are breaking instead of continuing while looping through all mom objects for hook related operations
1. https://github.com/openpbs/openpbs/blob/69d4e0258bd62541409305e3a0bd8b083dc52149/src/server/hook_func.c#L5072-L5075
2. https://github.com/openpbs/openpbs/blob/69d4e0258bd62541409305e3a0bd8b083dc52149/src/server/hook_func.c#L5161-L5164

subhasisb commented:
since we no longer have "proc" (we do not fork to send anymore), can we rename the variables with "proc" in them? For example sync_mom_hookfiles_proc_running could be renamed to something simple like "sync_mom_hookfiles_pending"?

Shrini-h commented:
I have identified few memory warnings from valgrind
[valgrind_warnings.txt](https://github.com/openpbs/openpbs/files/5127934/valgrind_warnings.txt)

Will be addressing them in upcoming commit

Shrini-h commented:
> since we no longer have "proc" (we do not fork to send anymore), can we rename the variables with "proc" in them? For example sync_mom_hookfiles_proc_running could be renamed to something simple like "sync_mom_hookfiles_pending"?

Agreed, I have renamed `sync_mom_hookfiles_proc_running` to `sync_mom_hookfiles_replies_pending` in https://github.com/openpbs/openpbs/pull/1992/commits/8f993df38b002f743619b9e003868ffc80d9a859



> I have identified few memory warnings from valgrind
> [valgrind_warnings.txt](https://github.com/openpbs/openpbs/files/5127934/valgrind_warnings.txt)
> 
> Will be addressing them in upcoming commit

I have addressed these valgrind warnings in https://github.com/openpbs/openpbs/pull/1992/commits/8f993df38b002f743619b9e003868ffc80d9a859

subhasisb commented:
looks good to me, will wait for final sign off from @bayucan

Shrini-h commented:
> It's all good to me too! Good work @Shrini-h !

Thank you @bayucan  for your guidance and approval !
